tapiriik/services/Hexoskin/hexoskin.py

```diff
@@ -0,0 +1,256 @@
+        logger.debug('Hexoskin - %s activities found, %s excluded. Activities' % ([x.ServiceData['ActivityID'] for x in activities], [x.ExternalActivityID for x in exclusions]))
+        return activities, exclusions
+
+
+    def DownloadActivity(self, serviceRecord, activity):
+        """Extract activity from Hexoskin"""
+        activityID = activity.ServiceData["ActivityID"]
+        logger.debug('Hexoskin - Extracting activity %s' % activityID)
+        headers = self._apiHeaders(serviceRecord)
+        headers.update({"Accept":"application/vnd.garmin.tcx+xml"})
+        range_tcx = requests.get(serverRoot + "api/range/%s/" % (str(activityID)), headers=headers)
+        TCXIO.Parse(range_tcx.content, activity)
+        activity.Notes = activity.Name
+        activity.Name = '%s - Hx' % (activity.Type)
+        return activity
+
+
+    def UploadActivity(self, serviceRecord, activity):
+        """Import data into Hexoskin using TCX format"""
+        tcx_data = TCXIO.Dump(activity)
+        headers = self._apiHeaders(serviceRecord)
+        headers.update({"Content-Type":"application/vnd.garmin.tcx+xml"})
+
+        range_tcx = requests.post(serverRoot + "api/import/", data=tcx_data.encode('utf-8'), headers=headers)
+        import_id = str(range_tcx.json()['resource_uri'])
+        # TODO In line below, fix the way the resource_uri is constructed when the /importfile/ fix is deployed
+        uploaded = False
+        process_start_time = datetime.now()
+        while uploaded is False:
+            time.sleep(1)
+            range_upload_status = requests.get(serverRoot + import_id, headers=headers).json()
+            if (datetime.now() - process_start_time).total_seconds() < 20:
+                if range_upload_status['results'] is not None and 'error' in range_upload_status['results']:
+                    err = range_upload_status['results']['error']
+                    logger.debug('Hexoskin - Import range failed, see error')
+                    raise APIException('Error uploading to Hexoskin: %s' % err)
+                elif range_upload_status['progress'] == 1 and range_upload_status['results']:
+                    for entry in range_upload_status['results']:
+                        if 'resource_uri' in entry.keys() and "range" in entry['resource_uri'] and entry['rank'] == 0:
+                            upload_id = entry['id']
+                            logger.debug('Hexoskin - Imported range %s' % upload_id)
+                            uploaded = True
+                            break
+            else:
+                raise APIException('Timeout uploading activity for import id %s' % import_id)  # when the '0' bug is fixed, change this
+        return upload_id
+
+
+    def DeleteCachedData(self, serviceRecord):
+        """No cached data"""
+        pass
+
+
+    def DeleteActivity(self, serviceRecord, uploadId):
+        """We would rather have users delete data from their dashboard instead of an automatic tool"""

```

Comment:
This is for the automatic rollback tool (used for recovering from significant issues with duplicates). Here, having an automated system remove only the activities tapiriik uploaded is much more reliable than the user trying to figure out which was the original and which was a duplicate, especially over 1,000s of activities.